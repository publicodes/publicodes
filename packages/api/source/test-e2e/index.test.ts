import Router from '@koa/router'
import chaiHttp from 'chai-http'
import Koa from 'koa'
import Engine from 'publicodes'
import { beforeAll, chai, describe, expect, it } from 'vitest'
import publicodesAPI from '../middleware/koa'

interface State extends Koa.DefaultState {}

interface Context extends Koa.DefaultContext {}

const app = new Koa<State, Context>()
const router = new Router<State, Context>()

const apiRoutes = publicodesAPI(() => new Engine('coucou: 0'))

router.use(apiRoutes)

app.use(router.routes())
app.use(router.allowedMethods())

const server = app.listen(3004)

chai.use(chaiHttp)

beforeAll(async () => {
	await new Promise((res, _rej) => {
		server.once('listening', () => res(null))
	})
})

describe('e2e koa middleware', () => {
	it('Test evaluate endpoint', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: '1 + 1' })
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(
			`
			{
			  "evaluate": [
			    {
			      "missingVariables": [],
			      "nodeValue": 2,
			      "traversedVariables": [],
			    },
			  ],
			  "situationError": null,
			}
		`
		)
	})

	it('Test evaluate endpoint with bad expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: '1+1' })
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchSnapshot()
	})

	it('Test evaluate endpoint with null expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: null })
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchSnapshot()
	})
	
	it('Test evaluate endpoint with empty expression', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({ expressions: [] })
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchSnapshot()
	})

	it('Test evaluate endpoint with bad situation', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({
					expressions: ['coucou = 2 * 21'],
					situation: 'coucou: 42',
				})
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": [],
			      "nodeValue": false,
			      "traversedVariables": [
			        "coucou",
			      ],
			    },
			  ],
			  "situationError": {
			    "message": "
			[ Erreur syntaxique ]
			➡️  Dans la règle \\"situation [0]\\"
			✖️  La référence 'c' est introuvable.
				Vérifiez que l'orthographe et l'espace de nom sont corrects
			    
			",
			  },
			}
		`)
	})

	it('Test evaluate endpoint with situation', async () => {
		await expect(
			chai
				.request(server)
				.post('/evaluate')
				.send({
					expressions: ['coucou = 2 * 21'],
					situation: { coucou: 42 },
				})
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(`
			{
			  "evaluate": [
			    {
			      "missingVariables": [],
			      "nodeValue": true,
			      "traversedVariables": [
			        "coucou",
			      ],
			    },
			  ],
			  "situationError": null,
			}
		`)
	})

	it('Test rules endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules')
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(`
			{
			  "coucou": {
			    "nodeKind": "rule",
			    "rawNode": {
			      "formule": "0",
			      "nom": "coucou",
			    },
			    "replacements": [],
			    "suggestions": {},
			    "title": "Coucou",
			  },
			}
		`)
	})

	it('Test rules/:id endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules/coucou')
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(`
			{
			  "nodeKind": "rule",
			  "rawNode": {
			    "formule": "0",
			    "nom": "coucou",
			  },
			  "suggestions": {},
			  "title": "Coucou",
			}
		`)
	})

	it('Test bad rules/:id endpoint', async () => {
		await expect(
			chai
				.request(server)
				.get('/rules/bad rule')
				.then((res) => JSON.parse(res.text))
		).resolves.toMatchInlineSnapshot(`
			{
			  "error": {
			    "message": "La règle 'bad rule' n'existe pas",
			  },
			}
		`)
	})
})
