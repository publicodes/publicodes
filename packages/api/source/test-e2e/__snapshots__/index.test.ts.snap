// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e koa middleware > Test evaluate endpoint with bad expression 2`] = `
{
  "evaluate": [
    {
      "error": {
        "message": "
[ Erreur syntaxique ]
➡️  Dans la règle \\"$EVALUATION\\"
✖️  \`1+1\` n'est pas une expression valide
    Syntax error at line 1 col 2:

1  1+1
    ^
Unexpected additionSubstraction token: \\"+\\". Instead, I was expecting to see one of the following:

A space token based on:
    number$ebnf$1$subexpression$1 →  ● %space
    number$ebnf$1 →  ● number$ebnf$1$subexpression$1
    number → %number ● number$ebnf$1 Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    Words →  ● %word
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A word token based on:
    WordOrKeyword →  ● %word
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A boolean token based on:
    WordOrKeyword →  ● %boolean
    Words →  ● WordOrKeyword Words$ebnf$1
    UnitNumerator →  ● Words UnitNumerator$ebnf$1
    Unit$ebnf$1 →  ● UnitNumerator
    Unit →  ● Unit$ebnf$1 Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
A space token based on:
    UnitDenominator$ebnf$1$subexpression$1 →  ● %space
    UnitDenominator$ebnf$1 →  ● UnitDenominator$ebnf$1$subexpression$1
    UnitDenominator →  ● UnitDenominator$ebnf$1 \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A \\"/\\" based on:
    UnitDenominator → UnitDenominator$ebnf$1 ● \\"/\\" Words
    Unit$ebnf$2 → Unit$ebnf$2 ● UnitDenominator
    Unit → Unit$ebnf$1 ● Unit$ebnf$2
    number → %number number$ebnf$1 ● Unit
    NumericTerminal →  ● number
    Parentheses →  ● NumericTerminal
    MultiplicationDivision →  ● Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    MultiplicationDivision → MultiplicationDivision ● %space %multiplicationDivision %space Parentheses
    AdditionSubstraction →  ● MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    AdditionSubstraction → AdditionSubstraction ● %space %additionSubstraction %space MultiplicationDivision
    NumericValue →  ● AdditionSubstraction
    main →  ● NumericValue
A space token based on:
    Comparison → Comparable ● %space %comparison %space Comparable
    main →  ● Comparison
",
      },
      "warnings": [],
    },
  ],
  "warnings": [],
}
`;

exports[`e2e koa middleware > Test evaluate endpoint with empty expression 2`] = `
[
  {
    "dataPath": ".expressions",
    "keyword": "minItems",
    "message": "should NOT have fewer than 1 items",
    "params": {
      "limit": 1,
    },
    "schemaPath": "#/properties/expressions/oneOf/0/minItems",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be number",
    "params": {
      "type": "number",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/1/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/2/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf",
  },
]
`;

exports[`e2e koa middleware > Test evaluate endpoint with null expression 2`] = `
[
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be array",
    "params": {
      "type": "array",
    },
    "schemaPath": "#/properties/expressions/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be number",
    "params": {
      "type": "number",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/0/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be string",
    "params": {
      "type": "string",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/1/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "type",
    "message": "should be object",
    "params": {
      "type": "object",
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf/2/type",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf/1/oneOf",
  },
  {
    "dataPath": ".expressions",
    "keyword": "oneOf",
    "message": "should match exactly one schema in oneOf",
    "params": {
      "passingSchemas": null,
    },
    "schemaPath": "#/properties/expressions/oneOf",
  },
]
`;
